cloudnew/ec2_client.go

```diff
@@ -0,0 +1,193 @@
+package cloudnew
+
+import (
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/evergreen-ci/evergreen/cloud"
+	"github.com/pkg/errors"
+)
+
+// AWSClient is a wrapper for aws-sdk-go so we can use a mock in testing.
+type AWSClient interface {
+	// Create a new aws-sdk-client or mock if one does not exist, otherwise no-op.
+	Create(*credentials.Credentials) error
+
+	// RunInstances is a wrapper for ec2.RunInstances.
+	RunInstances(*ec2.RunInstancesInput) (*ec2.Reservation, error)
+
+	// DescribeInstances is a wrapper for ec2.DescribeInstances.
+	DescribeInstances(*ec2.DescribeInstancesInput) (*ec2.DescribeInstancesOutput, error)
+
+	// CreateTags is a wrapper for ec2.CreateTags.
+	CreateTags(*ec2.CreateTagsInput) error
+
+	// TerminateInstances is a wrapper for ec2.TerminateInstances.
+	TerminateInstances(*ec2.TerminateInstancesInput) (*ec2.TerminateInstancesOutput, error)
+
+	// RequestSpotInstances is a wrapper for ec2.RequestSpotInstances.
+	RequestSpotInstances(*ec2.RequestSpotInstancesInput) (*ec2.RequestSpotInstancesOutput, error)
+
+	// DescribeSpotInstanceRequests is a wrapper for ec2.DescribeSpotInstanceRequests.
+	DescribeSpotInstanceRequests(*ec2.DescribeSpotInstanceRequestsInput) (*ec2.DescribeSpotInstanceRequestsOutput, error)
+
+	// CancelSpotInstanceRequests is a wrapper for ec2.CancelSpotInstanceRequests.
+	CancelSpotInstanceRequests(*ec2.CancelSpotInstanceRequestsInput) error
+}
+
+// AWSClientImpl wraps ec2.EC2.
+type AWSClientImpl struct {
+	session *session.Session
+	*ec2.EC2
+}
+
+// Create a new aws-sdk-client if one does not exist, otherwise no-op.
+func (c *AWSClientImpl) Create(creds *credentials.Credentials) error {
+	if creds == nil {
+		return errors.New("creds must not be nil")
+	}
+	if c.session == nil {
+		s, err := session.NewSession()
```

Comment:
is there a way to put one of our pooled http clients into this session? 

I want to avoid having a leak because of keep alive or some such. 

this might require putting in a close method to return the client to the pool. (alt: we could make this method take an http client.)

Reply:
done